core/src/main/java/org/apache/accumulo/core/util/tables/TableZooHelper.java

```diff
@@ -62,82 +55,81 @@ public TableZooHelper(ClientContext context) {
+        "No entry for this table Id found in table mappings");
   }
 
-  public String getTableName(TableId tableId) throws TableNotFoundException {
-    for (AccumuloTable systemTable : AccumuloTable.values()) {
-      if (systemTable.tableId().equals(tableId)) {
-        return systemTable.tableName();
+  private Map<String,String> loadQualifiedTableMapping(boolean reverse) {
+    final var builder = ImmutableMap.<String,String>builder();
+    for (NamespaceId namespaceId : context.getNamespaces().getIdToNameMap().keySet()) {
+      for (Map.Entry<TableId,String> entry : context.getTableMapping(namespaceId).getIdToNameMap()
+          .entrySet()) {
+        String fullyQualifiedName;
+        try {
+          fullyQualifiedName = TableNameUtil.qualified(entry.getValue(),
+              Namespaces.getNamespaceName(context, namespaceId));
+        } catch (NamespaceNotFoundException e) {
+          throw new RuntimeException(
+              "getNamespaceName() failed to find namespace for namespaceId: " + namespaceId, e);
+        }
+        if (reverse) {
+          builder.put(fullyQualifiedName, entry.getKey().canonical());
+        } else {
+          builder.put(entry.getKey().canonical(), fullyQualifiedName);
+        }
       }
     }
-    String tableName = getTableMap().getIdtoNameMap().get(tableId);
-    if (tableName == null) {
-      throw new TableNotFoundException(tableId.canonical(), null, null);
-    }
-    return tableName;
+    return builder.build();
   }
 
-  /**
-   * Get the TableMap from the cache. A new one will be populated when needed. Cache is cleared
-   * manually by calling {@link #clearTableListCache()}
-   */
-  public TableMap getTableMap() {
-    final ZooCache zc = context.getZooCache();
-    TableMap map = getCachedTableMap();
-    if (!map.isCurrent(zc)) {
-      instanceToMapCache.invalidateAll();
-      map = getCachedTableMap();
-    }
-    return map;
+  public Map<String,TableId> getQualifiedNameToIdMap() {
+    return loadQualifiedTableMapping(true).entrySet().stream()
+        .collect(Collectors.toMap(Map.Entry::getKey, e -> TableId.of(e.getValue())));
   }
 
-  private TableMap getCachedTableMap() {
-    return instanceToMapCache.get(this, k -> new TableMap(context));
+  public Map<TableId,String> getIdtoQualifiedNameMap() {
+    return loadQualifiedTableMapping(false).entrySet().stream()
+        .collect(Collectors.toMap(e -> TableId.of(e.getKey()), Map.Entry::getValue));
   }
 
   public boolean tableNodeExists(TableId tableId) {
-    ZooCache zc = context.getZooCache();
-    List<String> tableIds = zc.getChildren(Constants.ZTABLES);
-    return tableIds.contains(tableId.canonical());
+    for (NamespaceId namespaceId : context.getNamespaces().getIdToNameMap().keySet()) {
+      for (Map.Entry<TableId,String> entry : context.getTableMapping(namespaceId).getIdToNameMap()
+          .entrySet()) {
+        if (entry.getKey().equals(tableId)) {
+          return true;
+        }
+      }
```

Comment:
Could the scan of the be avoided by doing the following?

```suggestion
      if(context.getTableMapping(namespaceId).getIdToNameMap().containsKey(tableId)){
        return true;
      }
```